### PR TITLE
Fix editing performance in Widgets Customizer

### PR DESCRIPTION
**Pull request by kevin940726**
## Description

This fixes an important performance issue in the Widgets Customizer where editing any block will re-render the whole block editor. This branch is based on #30221 for easier review but should be re-target to trunk before merge.

The problem with the original approach is that every time we update any block we create a new instance of the `nextBlocks` array and set it back to the component's state via `setState`. This feels intuitive but apparently it breaks the optimization `useBlockSync` has. Turns out we need to make sure the `nextBlocks` instance we pass to the `<BlockEditorProvider>`'s `value` should be referential equal to the `nextBlocks` we received from the `onChange` callback. Or else all changes are marked as persistent and will re-render the whole block editor.

The solution is simply to call `setState` with the `nextBlocks` we receive from the callback whenever possible.

I also essentially refactored the code in `SidebarAdaptor` and `useSidebarBlockEditor` to use the customize API as the source of truth of all states, and simplify the logic a little bit. I added a few methods regarding undo/redo in #30400. They're currently unused yet though, but they should work. We can add the UI in another PR.

## How has this been tested?

1. Enable "Highlight updates" in React devtools.
2. Enable Widgets Customizer in the gutenberg experiments page.
3. Go to Appearance -> Customize -> Widgets.
4. Edit any block in any widget area.
5. Notice that only the block being edited is highlighted by React devtools most of the time.
6. And of course, it's way smoother than before.
7. Also try adding a new block, creating block on enter, or deleting a block. And notice the corresponding changes should reflect on the preview as expected.

## Screenshots 

https://user-images.githubusercontent.com/7753001/114140441-d677d180-9942-11eb-8285-dcec6520bbdc.mp4

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js [hunk 1]
```diff
@@ -170,24 +179,124 @@ export default class SidebarAdapter {
...
+		}
+
+		this._seek( this.historyIndex - 1 );
+	}
+	redo() {
+		if ( ! this.canRedo() ) {
+			return;
+		}
+
+		this._seek( this.historyIndex + 1 );
```
Comment (kevin940726) [inline packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js]: These are currently unused, but paving the way for the undo/redo feature in the future.

File: packages/customize-widgets/src/components/header/index.js [hunk 2]
```diff
@@ -45,4 +45,4 @@ function Header( { inserter, isInserterOpened, setIsInserterOpened } ) {
 	);
 }
 
-export default Header;
+export default memo( Header );
```
Comment (kevin940726) [inline packages/customize-widgets/src/components/header/index.js]: Memoing this allow us to prevent re-render it every time we update a block.

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-editor-provider.js [hunk 3]
```diff
@@ -23,6 +36,14 @@ export default function SidebarEditorProvider( {
 			settings={ settings }
 			useSubRegistry={ false }
 		>
+			<BlockEditorKeyboardShortcuts />
+
+			<Header
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/sidebar-editor-provider.js]: Why did you move `Header` from `SidebarBlockEditor` to `SidebarEditorProvider`? 

A `Provider` component should provide context and not render any elements. Can we move `Header` elsewhere or rename `SidebarEditorProvider`?

File: packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js [hunk 4]
```diff
@@ -88,175 +93,109 @@ function widgetToBlock( widget ) {
...
+			} );
+		} );
 	}, [ sidebar ] );
 
 	const onChangeBlocks = useCallback(
-		( _nextBlocks ) => {
-			ignoreIncoming.current = true;
+		( nextBlocks ) => {
+			setBlocks( ( prevBlocks ) => {
+				if ( isShallowEqual( prevBlocks, nextBlocks ) ) {
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js:150]: Is this ever true? I don't understand why `useBlockSync` would call `onChange` if nothing has changed.

File: packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js [hunk 5]
```diff
@@ -88,175 +93,109 @@ function widgetToBlock( widget ) {
...
+				const nextWidgets = nextBlocks.map( ( nextBlock ) => {
+					const widgetId = getWidgetId( nextBlock );
+
+					// Update existing widgets.
+					if ( widgetId && prevBlocksMap.has( widgetId ) ) {
+						const prevBlock = prevBlocksMap.get( widgetId );
+						const prevWidget = sidebar.getWidget( widgetId );
+
+						// Bail out updates by returning the previous widgets.
+						// Deep equality is necessary until the block editor's internals changes.
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js:170]: What do you mean by "until the block editor's internals change"? Is this ever going to happen?

File: packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js [hunk 6]
```diff
@@ -88,175 +93,109 @@ function widgetToBlock( widget ) {
...
+							return addWidgetIdToBlock(
+								nextBlock,
+								addedWidgetId
+							);
+						}
 
-				ignoreIncoming.current = false;
+						return nextBlock;
+					} );
+				}
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js]: We can combine the `filter` of `addedWidgetIds` and the `map` of `nextBlocks` here if we use `reduce`. That way there's only one array traversal. I'm not sure if it would actually make a difference to real world performance, though.

```js
const addedWidgetIds = sidebar.setWidgets( nextWidgets );

return nextBlocks.reduce(
	( updatedNextBlocks, nextBlock, index ) => {
		const addedWidgetId = addedWidgetIds[ index ];

		if ( addedWidgetId !== null ) {
			if ( updatedNextBlocks === nextBlocks ) {
				updatedNextBlocks = { ...nextBlocks };
			}
			updatedNextBlocks[ index ] = addWidgetIdToBlock(
				nextBlock,
				addedWidgetId
			);
		}

		return updatedNextBlocks;
	},
	nextBlocks
);
```

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js [hunk 7]
```diff
@@ -53,12 +48,25 @@ export default class SidebarAdapter {
...
+		this.historyIndex = 0;
+
 		this._handleSettingChange = this._handleSettingChange.bind( this );
 		this._handleAllSettingsChange = this._handleAllSettingsChange.bind(
 			this
 		);
+		this.canUndo = this.canUndo.bind( this );
+		this.canRedo = this.canRedo.bind( this );
+		this.undo = this.undo.bind( this );
+		this.redo = this.redo.bind( this );
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js:69]: Binding these shouldn't be necessary. It's only needed when we use a class member function as a callback.

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js [hunk 8]
```diff
@@ -53,12 +48,25 @@ export default class SidebarAdapter {
...
+		this.locked = false;
+		this.widgetsMap = new WeakMap();
 		this.subscribers = new Set();
 
+		this.history = [
+			this._getWidgetIds().map( ( widgetId ) =>
+				this.getWidget( widgetId )
+			),
+		];
+		this.historyIndex = 0;
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js:60]: These properties could be defined at the class level if you wanted to clean up the constructor a little. Up to you.

```js
class SidebarAdapter {
	locked = false;

	widgetsMap = new WeakMap();

	subscribers = new Set();

	history = [];

	historyIndex = 0;

	constructor( setting, allSettings ) {
		this.setting = setting;
		this.allSettings = allSettings;

		this.history = [
 			this._getWidgetIds().map( ( widgetId ) =>
 				this.getWidget( widgetId )
 			),
 		];

		this._handleSettingChange = this._handleSettingChange.bind( this );
 		this._handleAllSettingsChange = this._handleAllSettingsChange.bind( this );
	}
}
```

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js [hunk 9]
```diff
@@ -82,30 +90,47 @@ export default class SidebarAdapter {
...
 		for ( const callback of this.subscribers ) {
-			callback( event );
+			callback( ...args );
 		}
 	}
 
-	_handleSettingChange( newWidgetIds, oldWidgetIds ) {
-		const addedWidgetIds = difference( newWidgetIds, oldWidgetIds );
-		const removedWidgetIds = difference( oldWidgetIds, newWidgetIds );
+	_getWidgetIds() {
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js:100]: It's slightly confusing that `getWidgets()` and `_getWidgetIds()` are named similarly but the former returns "adapted" data and the returns data from the underlying setting. I'd consider not having `_getWidgetIds()` and instead using `this.setting.get()` everywhere. It's more verbose but, in my opinion, less ambiguous.

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js [hunk 10]
```diff
@@ -170,24 +179,124 @@ export default class SidebarAdapter {
...
+			number,
+			instance,
+		};
+
+		this.widgetsMap.set( instance, widget );
+
+		return widget;
+	}
+
+	_internalUpdateWidgets( nextWidgets ) {
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js]: I don't think we need to say `internal`. The `_` prefix makes that clear enough.

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js [hunk 11]
```diff
@@ -170,24 +179,124 @@ export default class SidebarAdapter {
...
+		const addedWidgetIds = this._internalUpdateWidgets( nextWidgets );
+
+		this._pushHistory();
+
+		return addedWidgetIds;
+	}
+
+	canUndo() {
+		return this.historyIndex > 0;
+	}
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js:273]: Let's add a line whitespace in between these methods for consistency.

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js [hunk 12]
```diff
@@ -170,24 +179,124 @@ export default class SidebarAdapter {
...
+		const settingId = widgetIdToSettingId( widgetId );
+		const setting = this.allSettings( settingId );
+
+		if ( ! setting ) {
+			return null;
+		}
+
+		const instance = setting.get();
+
+		if ( this.widgetsMap.has( instance ) ) {
```
Comment (noisysocks) [inline packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js]: Is `widgetsMap` only used for caching the return value of `getWidget()`? If so then I think let's name it `widgetsCache` or `cache`.

**Review by noisysocks — COMMENTED** [review]
This is fantastic! I tested this locally and it's a huge huge huge improvement. `useSidebarBlockEditor` is a lot easier to understand now, too. I left a few minor comments.

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-editor-provider.js (hunk 3 above)
Comment (kevin940726) [reply packages/customize-widgets/src/components/sidebar-block-editor/sidebar-editor-provider.js]: Yeah, I think we can move it out! It's from a previous experiment I forgot to refactor.

File: packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js (hunk 4 above)
Comment (kevin940726) [reply packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js:150]: It happened once in my testing, not sure why though 🤔.

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js (hunk 7 above)
Comment (kevin940726) [reply packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js:69]: It's just easier if any components/hooks want to access these methods without having to memoize/bind them again. We don't need these now though.

```js
const { canUndo, canRedo, undo, redo } = sidebar;
```

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js (hunk 9 above)
Comment (kevin940726) [reply packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js:100]: I tried that but I thought `this.setting.get()` isn't very clear what it would return.

File: packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js (hunk 8 above)
Comment (kevin940726) [reply packages/customize-widgets/src/components/sidebar-block-editor/sidebar-adapter.js:60]: It's a syntax error though since that we never enabled class properties in our babel settings 😅.

File: packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js (hunk 5 above)
Comment (kevin940726) [reply packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js:170]: It seems like we currently use `isEqual` under the hood in our reducer for every edit in the block editor package. Until we optimize it there, it's best to follow the same logic here and do a `isEqual` check for every edit as well. I have no idea if we ever have time to optimize it though 😅.

File: packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js (hunk 6 above)
Comment (kevin940726) [reply packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js]: I don't think there will be any difference since it's not the hot path and it's still O(N), but I like how it looks so I'll steal it 😆.

File: packages/customize-widgets/src/components/header/index.js (hunk 2 above)
Comment (kevin940726) [reply packages/customize-widgets/src/components/header/index.js]: We don't need this now as the `<Header>` component has been moved outside of the provider.

**Review by noisysocks — APPROVED** [review]

**Comment by kevin940726** [comment]
Shoot, I kept forgetting to change my base before rebasing and force pushing, sorry for the pings folks 😅.

File: packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js (hunk 4 above)
Comment (noahtallen) [reply packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js:150]: IIRC, `useBlockSync` works based on the [reference to the blocks array](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/block-editor/src/components/provider/use-block-sync.js#L188); I don't know if it goes through the array of blocks. It theoretically works because the `block-editor` store is supposed to make sure that references to a slice of blocks don't change unless the blocks actually did change. (To prevent re-renders and stuff.)

It might be worth changing [that line](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/block-editor/src/components/provider/use-block-sync.js#L188) to use `isShawllowEqual` as well. I think it may have been avoided to prevent looping over the entire array of blocks every time the subscription callback fires

Comment (kevin940726) [reply packages/customize-widgets/src/components/sidebar-block-editor/use-sidebar-block-editor.js:150]: > It might be worth changing [that line](https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/block-editor/src/components/provider/use-block-sync.js#L188) to use `isShawllowEqual` as well. I think it may have been avoided to prevent looping over the entire array of blocks every time the subscription callback fires

Yeah I think that would make sense! I believe it also means that I don't have to do the weird `if (nextBlocks === newBlocks) newBlocks = [...nextBlocks]` thing below.

Want to make an issue/PR?
